File: gross_to_net.m
function [net_income_s_h, net_income_s_w, net_income_m, net_income_m_unemp ] = gross_to_net(const_values, N_KIDS, wage_w, wage_h, t, age_index)
% the tax brackets and the deductions and exemptions starts at 1980 and
% ends at 2035. most of the sample turn 18 at 1980 (NLSY79)
row_number = t + age_index;
% row number on matrix 1980-2035. 

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
deductions_m = const_values.ded(row_number,2);
exemptions_m = const_values.ded(row_number,4) + const_values.ded(row_number,6)*N_KIDS;
deductions_s = const_values.ded(row_number,3); 
exemptions_s_w = const_values.ded(row_number,5) + const_values.ded(row_number,6)*N_KIDS;
exemptions_s_h = const_values.ded(row_number,5);

net_income_s_h = 0;net_income_s_w = 1;net_income_m = 0;net_income_m_unemp = 0;
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% CALCULATE NET INCOME FOR SINGLE WOMEN
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
if (wage_w>0)
    shift = 0; % single 2-22, married 23-42
    reduced_income_s_w = wage_w - deductions_s - exemptions_s_w;
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    %  CALCULATE INCOME TAX           %
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    tax = 0;
    if reduced_income_s_w > 0
        for i = 2:10
            if (reduced_income_s_w < const_values.tax(row_number,i+shift))
                tax = tax + (reduced_income_s_w - const_values.tax(row_number,(i-1)+shift))*const_values.tax(row_number,11+(i-1)+shift);
            break
            end
            tax = tax + (const_values.tax(row_number,i+shift) - const_values.tax(row_number,(i-1)+shift)) ...
                   * const_values.tax(row_number,11+(i-1)+shift);
        end
    end    
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    %  CALCULATE EICT                 %
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    EICT = 0;
 
    if (wage_w < const_values.ded(row_number,8+6*N_KIDS))               % first interval  credit rate
        EICT = wage_w*const_values.ded(row_number,7+6*N_KIDS);
        tax = 0;
    elseif  ((wage_w > const_values.ded(row_number,8+6*N_KIDS))...      % second (flat) interval - max EICT 
            && (wage_w < const_values.ded(row_number,11+6*N_KIDS)))
        EICT = const_values.ded(row_number,9+6*N_KIDS);
        tax = 0;
    elseif  ((wage_w > const_values.ded(row_number,11+6*N_KIDS))...     % third interval - phaseout rate
            && (wage_w < const_values.ded(row_number,12+6*N_KIDS)))
        EICT = wage_w*const_values.ded(row_number,10+6*N_KIDS);
        tax = 0;
    else
        EICT = 0;                                                       % income too high for EICT
    end    
    net_income_s_w = wage_w - tax + EICT; 
    
end
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% CALCULATE NET INCOME FOR SINGLE MEN
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
if (wage_h > 0)
    shift = 0; % single 2-22, married 23-42
    reduced_income_s_h =  wage_h- deductions_s-exemptions_s_h;
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    %  CALCULATE INCOME TAX           %
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    tax = 0;
    if reduced_income_s_h > 0
        for i = 2:10
            if (reduced_income_s_h < const_values.tax(row_number,i+shift))
                tax = tax + (reduced_income_s_h - const_values.tax(row_number,(i-1)+shift))*const_values.tax(row_number,11+(i-1)+shift);
            break
            end
            tax = tax + (const_values.tax(row_number,i+shift) - const_values.tax(row_number,(i-1)+shift)) ...
                   * const_values.tax(row_number,11+(i-1)+shift);
        end
    end    
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    %  CALCULATE EICT                 %
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    EICT = 0;
    N_KIDS_SINGLE = 0;
    if (wage_h < const_values.ded(row_number,8+6*N_KIDS_SINGLE))               % first interval  credit rate
        EICT = wage_h*const_values.ded(row_number,7+6*N_KIDS_SINGLE);
        tax = 0;
    elseif  ((wage_h > const_values.ded(row_number,8+6*N_KIDS_SINGLE))...      % second (flat) interval - max EICT 
            && (wage_h < const_values.ded(row_number,11+6*N_KIDS_SINGLE)))
        EICT = const_values.ded(row_number,9+6*N_KIDS_SINGLE);
        tax = 0;
    elseif  ((wage_h > const_values.ded(row_number,11+6*N_KIDS_SINGLE))...     % third interval - phaseout rate
            && (wage_h < const_values.ded(row_number,12+6*N_KIDS_SINGLE)))
        EICT = wage_h*const_values.ded(row_number,10+6*N_KIDS_SINGLE);
        tax = 0;
    else
        EICT = 0;                                                   % income too high for EICT
    end    
    net_income_s_h = wage_h - tax + EICT; 
    
end
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%      NET SALARY FOR MARRIED COUPLE IF WOMEN CHOOSE TO WORK 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
if (wage_h > 0 )
    shift = 21; % single 2-22, married 23-42
    reduced_income_m = wage_h + wage_w  - deductions_m-exemptions_m;
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    %  CALCULATE INCOME TAX           %
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    tax = 0;
    if reduced_income_m > 0
        for i = 2:10
            if (reduced_income_m < const_values.tax(row_number,i+shift))
                tax = tax + (reduced_income_m - const_values.tax(row_number,(i-1)+shift))*const_values.tax(row_number,11+(i-1)+shift);
            break
            end
            tax = tax + (const_values.tax(row_number,i+shift) - const_values.tax(row_number,(i-1)+shift)) ...
                   * const_values.tax(row_number,11+(i-1)+shift);
        end
    end    
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    %  CALCULATE EICT                 %
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    EICT = 0;
    tot_inc = wage_h + wage_w;
    if (tot_inc < const_values.ded(row_number,8+6*N_KIDS))             % first interval  credit rate
        EICT = tot_inc*const_values.ded(row_number,7+6*N_KIDS);
        tax = 0;
    elseif  ((tot_inc > const_values.ded(row_number,8+6*N_KIDS))...    % second (flat) interval - max EICT 
            && (tot_inc < const_values.ded(row_number,11+6*N_KIDS)))
        EICT = const_values.ded(row_number,9+6*N_KIDS);
        tax = 0;
    elseif  ((tot_inc > const_values.ded(row_number,11+6*N_KIDS))...   % third interval - phaseout rate
            && (tot_inc < const_values.ded(row_number,12+6*N_KIDS)))
        EICT = tot_inc*const_values.ded(row_number,10+6*N_KIDS);
        tax = 0;
    else
        EICT = 0;                                           % income too high for EICT
    end    
    net_income_m = tot_inc - tax + EICT;
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% NET SALARY IF MARRIED AND WOMEN CHOOSES NOT TO WORK
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    shift = 21; % single 2-22, married 23-42
    reduced_income_m_unemp = wage_h - deductions_m-exemptions_m;
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    %  CALCULATE INCOME TAX           %
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    tax = 0;
    if reduced_income_m_unemp > 0
        for i=2:10
            if (reduced_income_m_unemp < const_values.tax(row_number,i+shift))
                tax = tax + (reduced_income_m_unemp - const_values.tax(row_number,(i-1)+shift))*const_values.tax(row_number,11+(i-1)+shift);
            break
            end
            tax = tax + (const_values.tax(row_number,i+shift) - const_values.tax(row_number,(i-1)+shift)) ...
                   * const_values.tax(row_number,11+(i-1)+shift);
        end
    end    
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    %  CALCULATE EICT if women married and unemployed               %
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    EICT = 0;
    tot_inc = wage_h;
    if (tot_inc < const_values.ded(row_number,8+6*N_KIDS))              % first interval  credit rate
        EICT = tot_inc*const_values.ded(row_number,7+6*N_KIDS);
        tax = 0;
    elseif  ((tot_inc > const_values.ded(row_number,8+6*N_KIDS))...     % second (flat) interval - max EICT 
            && (tot_inc < const_values.ded(row_number,11+6*N_KIDS)))
        EICT = const_values.ded(row_number,9+6*N_KIDS);
        tax = 0;
    elseif  ((tot_inc > const_values.ded(row_number,11+6*N_KIDS))...    % third interval - phaseout rate
            && (tot_inc < const_values.ded(row_number,12+6*N_KIDS)))
        EICT = tot_inc*const_values.ded(row_number,10+6*N_KIDS);
        tax = 0;
    else
        EICT = 0;                                                       % income too high for EICT
    end    
    net_income_m_unemp =  tot_inc - tax + EICT;
end
    

File: c++/gross_to_net.cpp
#include "gross_to_net.h"
#include "parameters.h"
#include <iostream>

// tax matrix
// -----+-------------------+------------------------------------------------------------------------------------------------------------------------------+----------
//      | single                                                                                                                                           | married 
// -----+-------------------------------------------------------------------+------------------------------------------------------------------------------+----------
// year | br1 | br2 | br3 | br4 | br5 | br6 | br7 | br8 | br9 | br10 | br11 | %br1 | %br2 | %br3 | %br4 | %br5 | %br6 | %br7 | %br8 | %br9 | %br10 | %br11 | ...
// -----+-------------------------------------------------------------------+------------------------------------------------------------------------------+----------

const auto TAX_PERCENT_OFFSET = 11;

// deduction matrix
// -----+-------------------------------------------------------------+--------------------------------------------+--------------------------------------------+--------
//      |                                                             | 0 kids                                     | 1 kid                                      | 2 kids
// -----+-------------------------------------------------------------+--------------------------------------------+--------------------------------------------+--------
// year | ded married | ded single | ex married | ex single | ex kids | int1% | int1 | int2# | int3% | int2 | int3 | int1% | int1 | int2# | int3% | int2 | int3 | ...
// -----+-------------------------------------------------------------+--------------------------------------------+--------------------------------------------+--------

const auto DED_OFFSET = 6U;
const auto DED_KIDS_OFFSET = 6U;
const auto DED_INTERVAL1_OFFSET = DED_OFFSET + 1;
const auto DED_INTERVAL2_OFFSET = DED_OFFSET + 4;
const auto DED_INTERVAL3_OFFSET = DED_OFFSET + 5;

double calculate_tax(const Parameters &p, double reduced_income, unsigned row_number) {
    double tax = 0.0;
    if (reduced_income > 0) {
        for (auto i = 2; i <= 11; ++i) {
            const auto lower_bracket = p.tax[row_number][i-1];
            const auto upper_bracket = p.tax[row_number][i];
            const auto percent = p.tax[row_number][i-1+TAX_PERCENT_OFFSET];
            if (reduced_income <= upper_bracket) {
                tax += (reduced_income - lower_bracket)*percent;
                break;
            }
            tax += (upper_bracket - lower_bracket)*percent;
        }
    }
    return tax;
}

double calculate_eict(const Parameters &p, double wage, unsigned row_number, unsigned kids) {
    double EICT = 0.0;

    const auto kids_offset = DED_KIDS_OFFSET*kids;
    if (wage < p.ded[row_number][DED_INTERVAL1_OFFSET+kids_offset]) {
        // first interval  credit rate
        EICT = wage*p.ded[row_number][DED_INTERVAL1_OFFSET-1+kids_offset];
    } else if (wage < p.ded[row_number][DED_INTERVAL2_OFFSET+kids_offset]) {
        // second (flat) interval - max EICT
        EICT = p.ded[row_number][DED_INTERVAL2_OFFSET-2+kids_offset]; 
    } else if (wage < p.ded[row_number][DED_INTERVAL3_OFFSET+kids_offset]) {
        EICT = wage*p.ded[row_number][DED_INTERVAL3_OFFSET-2+kids_offset];
    }

    return EICT;
}

// similar handling for husband and wife in case of singles
double gross_to_net_single(const Parameters &p, unsigned row_number, unsigned kids, double wage, double exemptions, double deductions) {
    if (wage > 0.0) {
        const auto reduced_income = wage - deductions - exemptions;
        double tax = 0.0;
        const auto EICT = calculate_eict(p, wage, row_number, kids);
        if (EICT == 0.0) {
            tax = calculate_tax(p, reduced_income, row_number);
        }

        return wage - tax + EICT;   
    }
    return 0.0;
}

double gross_to_net_married(const Parameters &p, unsigned row_number, unsigned kids, double wage_h, double wage_w, double exemptions, double deductions) {
    if (wage_h > 0.0) {
        const auto reduced_income = wage_h + wage_w - deductions - exemptions;
        double tax = 0.0;
        const auto tot_income = wage_h + wage_w;
        const auto EICT = calculate_eict(p, tot_income, row_number, kids);
        if (EICT == 0.0) {
            tax = calculate_tax(p, reduced_income, row_number);
        }
        return tot_income - tax + EICT;   
    }
    return 0.0;
}

NetIncome gross_to_net(const Parameters &p, unsigned kids, double wage_w, double wage_h, unsigned t, unsigned age_index) {
    // the tax brackets and the deductions and exemptions starts at 1980 and
    // ends at 2035. most of the sample turn 18 at 1980 (NLSY79)
    const auto row_number = t + age_index; // row number on matrix 1980-2035. 

    const auto deductions_m = p.ded[row_number][1];
    const auto deductions_s = p.ded[row_number][2];
    const auto exemptions_m = p.ded[row_number][3] + p.ded[row_number][5]*kids;
    const auto exemptions_s_w = p.ded[row_number][4] + p.ded[row_number][5]*kids;
    const auto exemptions_s_h = p.ded[row_number][4];

    NetIncome result;

    // CALCULATE NET INCOME FOR SINGLE WOMEN
    result.net_income_s_w = gross_to_net_single(p, row_number, kids, wage_w, exemptions_s_w, deductions_s);

    // CALCULATE NET INCOME FOR SINGLE MEN
    result.net_income_s_h = gross_to_net_single(p, row_number, 0 /*no kids*/, wage_h, exemptions_s_h, deductions_s); 

    // CALCULATE NET INCOME FOR MARRIED COUPLE
    result.net_income_m = gross_to_net_married(p, row_number, kids, wage_h, wage_w, exemptions_m, deductions_m);
    result.net_income_m_unemp = gross_to_net_married(p, row_number, kids, wage_h, 0.0 /*no wage_w*/, exemptions_m, deductions_m);

    return result;
}

